<a href="https://colab.research.google.com/github/kimberlyydanij/python_demo/blob/master/python_04_function.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
###function

def proc():
  pass # pass는 구현을 안할 떄 사용한다.
       # 아무것도 안 쓰면 에러가 발생한다.
       

In [1]:
def display() :
  pass

In [3]:
display()

In [5]:
print(display())

None


#### callbale
- 이름뒤에 괄호()를 붙이는 것을 call(호출) 이라고 한다.
- 이름뒤에 괄호()를 붙일 수 있는 것을 callable이라고 한다.
- 함수는 사용할 때 이름 뒤에 괄호를 붙인다.
- python에서 callable인 것이 3가지 유형이 있다.
  (1) function
  (2) clasS(init가 정의되었을 경우)
  (3) object(call가 정의되었을 경우)
  

In [6]:
a = 3
a()

TypeError: ignored

In [9]:
list = 3
list()

TypeError: ignored

In [10]:
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

In [11]:
list
set

set

In [12]:
list = 3
del list
list()

[]

In [16]:
def prn():
  return 1

In [17]:
prn()

1

In [18]:
# 괄호를 붙이지 않으면 함수식이다
prn

<function __main__.prn()>

In [20]:
# 내장함수 callable은 객체가()을 붙여 사용할 수 있는지를 True/False로 알려준다.
callable(prn)

True

In [22]:
a = 3
callable(a)

False

### 매개변수(parameter) 인자(argument)
1. positional
2. keyword
3. position & keyword
4. position only
5. keyword only
6. 
7. 가변 keyword

1. positional
   함수 호출 : 선언할 때 파라미터의 순서와 호출할 때 인자의 입력 순서를 그대로 매칭한다.

In [23]:
def proc(a,b):
  return a,b

In [24]:
proc(2,4)

(2, 4)

2. keyword 
   keyword 방식을 이용하면 순서와 관계없이 파라미터 이름에 따라 값이 대입된다.

In [25]:
proc(b=4, a=2)

(2, 4)

3. position & keyword 혼합

In [26]:
def proc(a,b,c):
  return a,b,c


In [27]:
proc(2, c=5, b=3)

(2, 3, 5)

In [28]:
# 한번 keyword 방식을 사용하면, 그 다음부터는 keyword방법만 사용할 수 있다.
proc(b=5, a=4, 8)

SyntaxError: ignored

In [29]:
#python 버전확인
import sys
sys.version

'3.8.15 (default, Oct 12 2022, 19:14:39) \n[GCC 7.5.0]'

4. position only
position only는 함수를 사용할 때 position 방식으로만 인자를 쓸 수 있도록 강제하는 방법이다.
/ 의 앞에 오는 인자들은 position 방식으로 입력할 수 있지만, keyword방식으로만 입력할 수 없다. position only함수를 선언하는 것은 phthon 3.8부터 지원한다.
이전 버전에서는 사전 정의된 position only함수는 있지만, 사용자가 만들 수는 없었다.